edge-hub/core/src/Microsoft.Azure.Devices.Routing.Core/checkpointers/Checkpointer.cs

```diff
@@ -68,9 +68,20 @@ public static async Task<Checkpointer> CreateAsync(string id, ICheckpointStore s
             return checkpointer;
         }
 
-        public void Propose(IMessage message)
+        public void Propose(IMessage message, Option<ulong> totalMessages)
         {
             this.Proposed = Math.Max(message.Offset, this.Proposed);
+
+            if (totalMessages.HasValue)
+            {
+                var messageCount = (ulong)(this.Proposed - this.Offset);
+                var count = totalMessages.GetOrElse(ulong.MaxValue);
+                if (messageCount > count)
+                {
+                    this.Offset = this.Proposed - (long)count;
```

Comment:
What is the logic here? This seems unrelated to the metric.... Are you saying the offset is incorrect in EdgeHub?

Reply:
This is what I mean by adjusting the offset. Without TTL these values should be exactly the same, but with TTL there can be a discrepancy.

Example 1: No TTL and we are processing normally
We have 100 messages proposed and say 20 processed. This means that the count should be 80. 30 minutes have not passed so the totalMessages are still 100. We want to show 80.

Example 2: No TTL and we are offline
Same as above, we should always show what proposed - offset equals

Example 3: TTL and we process normally
Say we have 100 messages and 20 have been processed. Let's pretend we hit the 30 min mark and remove 40 messages due to TTL. This means we really only have 60 messages and not 80. So, we should adjust offset to reflect this fact.

Example 4: TTL and offline
At 30 min mark for cleanup we should only have TTL * (number of messages / second). So, if TTL is 300 seconds and we send 1 message every 5 seconds then there can only be a max of 60 messages. So, if we have say 200 messages proposed and none processed; at the 30 min mark we should adjust the offset to reflect we only have 60 messages that can possibly be processed.

Example 5: Come back from crash
Much the same as above examples, after a crash we want to make sure we only count the messages that are actually there. Currently, the checkpointer takes the message offset as the way of determining how many messages. For details on this, if we crashed after processing 100 messages then came up and got a new message, that message would be marked as 101 and we would have 1 message left to process. If we instead had processed 200 and had an offset of 100, but due to TTL eliminated another 40 messages, then we shouldn't say there are 100 messages to process (queue length) but instead we should say 60.